Products/CMFPlone/recyclebin.py

```diff
@@ -0,0 +1,584 @@
+from datetime import datetime
+from datetime import timedelta
+from persistent.mapping import PersistentMapping
+from plone.registry.interfaces import IRegistry
+from Products.CMFPlone.controlpanel.browser.recyclebin import (
+    IRecycleBinControlPanelSettings,
+)
+from Products.CMFPlone.interfaces.recyclebin import IRecycleBin
+from zope.annotation.interfaces import IAnnotations
+from zope.component import getUtility
+from zope.component.hooks import getSite
+from zope.interface import implementer
+
+import logging
+import uuid
+
+
+logger = logging.getLogger("Products.CMFPlone.RecycleBin")
+
+ANNOTATION_KEY = "Products.CMFPlone.RecycleBin"
+
+
+@implementer(IRecycleBin)
+class RecycleBin:
+    """Stores deleted content items"""
+
+    def __init__(self, context=None):
+        """Initialize the recycle bin for a site
+
+        Args:
+            context: The Plone site object (optional when used as a utility)
+        """
+        self.context = context
+        # When used as a utility without context, we'll get the context on demand
+
+    def _get_context(self):
+        """Get the context (Plone site) if not already available"""
+        if self.context is None:
+            self.context = getSite()
+        return self.context
+
+    def _get_storage(self):
+        """Get the storage for recycled items"""
+        context = self._get_context()
+        annotations = IAnnotations(context)
+
+        if ANNOTATION_KEY not in annotations:
+            annotations[ANNOTATION_KEY] = PersistentMapping()
```

Comment:
Don't use PersistentMapping here. It stores the entire mapping in a single persistent object, so it won't perform well if many items are in the recycle bin.

A OOBTree would be better.

But really you should probably create a new persistent RecyleBinStorage class, which has the BTree as an attribute. That gives more flexibility if we need to add other structures later (such as other BTrees to index the deleted items and access them efficiently).